File: samples/ts/src/GraphMapNodes.ts
// UI settings
// Expression: gmap
//   Type: Map
//     Layout: Graph
//     Nodes: @mapCustomNodes
//     Properties: @mapCustomValue (optional)
//     Save: Object Attributes
//   Type: GraphMapNode
//     Layout: Graph
//     Nodes: @nodeCustomNodes
//     Edges: @nodeCustomEdges
//     Properties: @nodeCustomValue (optional)
//     Save: Object Type Attributes
class GraphMapNodes {

    public static runMain() {
        new GraphMapNodes().start();
    }

    public start() {
        const gmap: Map<GraphMapNode<string>, GraphMapNode<string>[]> = new Map<GraphMapNode<string>, GraphMapNode<string>[]>();
        const gedges: Map<string, number> = new Map<string, number>(); // edges 
        Extractor.gedges = gedges;

        // nodes
        const node1: GraphMapNode<string> = new GraphMapNode<string>("1");
        const node2: GraphMapNode<string> = new GraphMapNode<string>("2");
        const node3: GraphMapNode<string> = new GraphMapNode<string>("3");
        const node4: GraphMapNode<string> = new GraphMapNode<string>("4");
        
        gmap.set(node1, [node2, node3]);
        gmap.set(node2, [node3, node4]);
        gmap.set(node3, []);
        gmap.set(node4, []);

        // edges
        gedges.set(node1.getValue() + "," + node2.getValue(), 100);
        gedges.set(node1.getValue() + "," + node3.getValue(), 200);
        gedges.set(node2.getValue() + "," + node3.getValue(), 300);
        gedges.set(node2.getValue() + "," + node4.getValue(), 400);

        console.log("done");
    }
}

export class GraphMapNode<T> {

    private value: T;

    public getValue(): T {
        return this.value;
    }

    public constructor(value: T) {
        this.value = value;
    }

    public toString() {
        return this.value;
    }
}


// custom extractor
export class Extractor {
    static gedges: Map<string, number>;

    // register the custom attributes to extract
    // you can select these from the ui 
    // instead of modifying your objects
    public static registerAttrs(): [string, string[]][] {
        return [
            ["Map", ["mapCustomNodes", "mapCustomEdges", "mapCustomValue"]],
            ["GraphMapNode", ["nodeCustomNodes", "nodeCustomEdges", "nodeCustomValue"]],
        ];
    }

    public static extract_mapCustomNodes(
        obj: object,
        root: Map<GraphMapNode<string>, GraphMapNode<string>[]>
    ): GraphMapNode<string>[] {
        const nodes: GraphMapNode<string>[] = [];
        for (const key of root.keys())
            nodes.push(key);
        return nodes;
    }

    public static extract_mapCustomEdges(
        obj: Map<GraphMapNode<string>, GraphMapNode<string>[]>,
        root: Map<GraphMapNode<string>, GraphMapNode<string>[]>
    ): null {
        return null; // return null to hide edges
    }

    public static extract_mapCustomValue(
        obj: Map<GraphMapNode<string>, GraphMapNode<string>[]>,
        _root: Map<GraphMapNode<string>, GraphMapNode<string>[]>
    ): string[] {
        let sb = "";
        for (const key of obj.keys()) {
            if (sb.length > 0)
                sb += ",";
            sb += key.getValue();
        }
        return [sb];
    }

    public static extract_nodeCustomNodes(
        obj: GraphMapNode<string>,
        root: Map<GraphMapNode<string>, GraphMapNode<string>[]>
    ): GraphMapNode<string>[] {
        return root.get(obj) || [];
    }

    public static extract_nodeCustomEdges(
        obj: GraphMapNode<string>,
        root: Map<GraphMapNode<string>, GraphMapNode<string>[]>
    ): string[] {
        const edges: string[] = [];
        for (const child of root.get(obj) || []) {
            const edgeKey = obj.getValue() + "," + child.getValue();
            const edgeValue: number | undefined = Extractor.gedges.get(edgeKey);
            if (edgeValue)
                edges.push(String(edgeValue));
            else
                edges.push("");
        }
        return edges;
    }

    public static extract_nodeCustomValue(
        obj: GraphMapNode<string>,
        _root: Map<GraphMapNode<string>, GraphMapNode<string>[]>
    ): string[] {
        return [String(obj.getValue())];
    }
}

GraphMapNodes.runMain();

File: samples/java/src/app/GraphMapNodes.java
package app;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

// UI settings
// Expression: gmap
//   Type: HashMap
//     Layout: Graph
//     Nodes: @mapCustomNodes
//     Properties: @mapCustomValue (optional)
//     Save: Object Attributes
//   Type: GraphMapNodes.GraphMapNode
//     Layout: Graph
//     Nodes: @nodeCustomNodes
//     Edges: @nodeCustomEdges
//     Properties: @nodeCustomValue (optional)
//     Save: Object Type Attributes
public class GraphMapNodes {

    public static void main(String[] args) {

        // graph with hashmap node
        HashMap<GraphMapNode<String>, List<GraphMapNode<String>>> gmap = new HashMap<>();
        HashMap<String, Integer> gedges = new HashMap<>();
        Extractor.gedges = gedges;

        GraphMapNode<String> node1 = new GraphMapNode<String>("1");
        GraphMapNode<String> node2 = new GraphMapNode<String>("2");
        GraphMapNode<String> node3 = new GraphMapNode<String>("3");
        GraphMapNode<String> node4 = new GraphMapNode<String>("4");

        gmap.put(node1, new ArrayList<>(List.of(node2, node3)));
        gmap.put(node2, new ArrayList<>(List.of(node3, node4)));
        gmap.put(node3, new ArrayList<>());
        gmap.put(node4, new ArrayList<>());

        // edges
        gedges.put(node1.getValue() + "," + node2.getValue(), 100);
        gedges.put(node1.getValue() + "," + node3.getValue(), 200);
        gedges.put(node2.getValue() + "," + node3.getValue(), 300);
        gedges.put(node2.getValue() + "," + node4.getValue(), 400);

        System.out.println("done");
    }

    public static class GraphMapNode<T> {

        protected T value;

        public T getValue() {
            return value;
        }

        public GraphMapNode(T value) {
            this.value = value;
        }
    }

    // custom extractor
    static class Extractor {
        static HashMap<String, Integer> gedges;

        // register the custom attributes to extract
        // you can select these from the ui
        // instead of modifying your objects
        public static Object[] registerAttrs() {
            return new Object[] {
                    new Object[] { "HashMap",
                            new String[] { "mapCustomNodes", "mapCustomEdges", "mapCustomValue" } },
                    new Object[] { "GraphMapNodes.GraphMapNode",
                            new String[] { "nodeCustomNodes", "nodeCustomEdges", "nodeCustomValue" } }
            };
        }

        public static GraphMapNode<String>[] extract_mapCustomNodes(
                HashMap<GraphMapNode<String>, List<GraphMapNode<String>>> obj,
                HashMap<GraphMapNode<String>, List<GraphMapNode<String>>> root) {
            List<GraphMapNode<String>> nodes = new ArrayList<GraphMapNode<String>>();
            if (root.keySet().size() > 0)
                nodes.addAll(root.keySet());
            return nodes.toArray(new GraphMapNode[0]);
        }

        public static String[] extract_mapCustomEdges(
                HashMap<GraphMapNode<String>, List<GraphMapNode<String>>> obj,
                HashMap<GraphMapNode<String>, List<GraphMapNode<String>>> root) {
            return null; // return null to hide edges
        }

        public static String[] extract_mapCustomValue(
                HashMap<GraphMapNode<String>, List<GraphMapNode<String>>> obj,
                HashMap<GraphMapNode<String>, List<GraphMapNode<String>>> root) {
            StringBuilder sb = new StringBuilder();
            for (GraphMapNode<String> key : root.keySet()) {
                sb.append(key.getValue());
                sb.append(",");
            }
            return new String[] { sb.toString() };
        }

        public static GraphMapNode<String>[] extract_nodeCustomNodes(
                GraphMapNode<String> obj,
                HashMap<GraphMapNode<String>, List<GraphMapNode<String>>> root) {
            List<GraphMapNode<String>> nodes = new ArrayList<GraphMapNode<String>>();
            if (root.containsKey(obj))
                nodes.addAll(root.get(obj));
            return nodes.toArray(new GraphMapNode[0]);
        }

        public static String[] extract_nodeCustomEdges(
                GraphMapNode<String> obj,
                HashMap<GraphMapNode<String>, List<GraphMapNode<String>>> root) {
            List<String> edges = new ArrayList<String>();
            if (root.containsKey(obj)) {
                for (GraphMapNode<String> child : root.get(obj)) {
                    String edgeKey = obj.getValue() + "," + child.getValue();
                    if (Extractor.gedges.containsKey(edgeKey)) {
                        int edgeValue = Extractor.gedges.get(edgeKey);
                        edges.add(String.valueOf(edgeValue));
                    } else {
                        edges.add("");
                    }
                }
            }
            return edges.toArray(new String[0]);
        }

        public static String[] extract_nodeCustomValue(
                GraphMapNode<String> obj,
                HashMap<GraphMapNode<String>, List<GraphMapNode<String>>> root) {
            return new String[] { obj.getValue().toString() };
        }
    }
}